<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="David Brocks" />

<meta name="date" content="2019-06-10" />

<title>Repexpr vignette</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Repexpr vignette</h1>
<h4 class="author">David Brocks</h4>
<h4 class="date">2019-06-10</h4>



<div id="prerequisites" class="section level1">
<h1>Prerequisites</h1>
<div id="genome-assembly-optional" class="section level2">
<h2>Genome assembly (optional)</h2>
<p>Repexpr can compute family-wise multiple sequence alignments of all repeat sequences in the genome to improve read mapping onto a consensus model. To do so, we require the genome assembly of our organism of choice stored as a <em>BSgenome</em> object. You can retrieve the full list of supported genomes by typing <code>BSgenome::available.genomes()</code> or create a custom <em>BSgenome</em> object following the <a href="https://www.bioconductor.org/packages//2.7/bioc/vignettes/BSgenome/inst/doc/BSgenomeForge.pdf">instructions</a>.</p>
<p>Since we are working with expression data from human cancer cell lines and mouse embryos, we will first install the UCSC hg38 and mm10 assemblies.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;BSgenome.Hsapiens.UCSC.hg38&quot;</span>)
BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;BSgenome.Mmusculus.UCSC.mm10&quot;</span>)</code></pre></div>
<p><strong>Important</strong> <strong>note</strong>: Do not use <strong>repeat-masked</strong> <em>BSgenome</em> objects (contain ‘masked’ suffix, e.g. <em>BSgenome.Hsapiens.UCSC.hg38.masked</em>)!</p>
</div>
<div id="mafft-multiple-alignment-program-optional" class="section level2">
<h2>MAFFT multiple alignment program (optional)</h2>
<p>To compute family-wise multiple sequence alignments (used to improve mapping of read/UMI signal along consensus TE models), Repexpr utilizes the <a href="https://mafft.cbrc.jp/alignment/software/">MAFFT multiple alignment program</a>. To use this feature, make sure mafft is in your command PATH.</p>
</div>
<div id="repeat-annotation-required" class="section level2">
<h2>Repeat annotation (required)</h2>
<p>A convient solution is to download transposon coordinates from the <a href="http://www.repeatmasker.org/genomicDatasets/RMGenomicDatasets.html">Repeatmasker homepage</a> or the <a href="http://www.dfam.org/releases/Dfam_3.0/annotations/">DFAM database</a> for your genome and assembly of choice. You can import the Repeatmasker <em>fa.out.gz</em> or DFAM <em>dfam.hits.gz</em> files using the Repexpr <code>importRMSK</code> and <code>importDFAM</code> functions, respectively. Another option is to provide custom annotation as long as it provides the basic information about chromosome, start, end, strand, repname (family identifier), and id_unique (unique locus identifier). In this tutorial, we will show you how to import such information using the provided example datasets.</p>
</div>
<div id="gene-model-annotation-required" class="section level2">
<h2>Gene model annotation (required)</h2>
<p>In addition to TE expression counts, Repexpr also quantifies genic expression levels using common gene interval and annotation formats (e.g. gtf). In this tutorial, we will use the <a href="https://www.gencodegenes.org/">Gencode</a> comprehensive gene annotation on the human reference chromosomes only. Other annotation ressources are currently untested and should be used with caution.</p>
</div>
<div id="readumi-alignments" class="section level2">
<h2>Read/UMI alignments</h2>
<p>Repexpr requires read alignment coordinates stored in <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">BAM format</a> as input, which are routinely generated during most common scRNA-seq workflows, including <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">10x’ Cellranger pipeline</a>. BAM inputs should be duplicate removed, position sorted, and indexed. Chunking BAM inputs (e.g. by chromosome) can accelerate the import into your R environment using the <code>importBAM</code> function. <strong>Important:</strong> Repexpr assumes the cell barcode is either stored as CB tag or BAM input files are seperated per cell. Other formats are currently not supported (e.g. cell barcode in read name).</p>
</div>
</div>
<div id="workflow-human-10x-scrna-seq-dataset-5" class="section level1">
<h1>Workflow human 10x scRNA-seq dataset (5’)</h1>
<p>In this tutorial, we are going to utilize 5’ scRNA-seq data on epigenetically de-repressed cancer cell lines to quantify transposable element (TE) expression levels at single-cell and locus resolution. Following the workflow, you’ll learn the specifics of Repexpr to adapt it to your single-cell dataset.</p>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>We start the workflow by loading Repexpr and the human hg38 <em>BSgenome</em> object into our R environment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw">load_all</span>(<span class="st">'/net/mraid14/export/tgdata/users/davidbr/src/Repexpr/'</span>)

<span class="co"># adjust to your genome of interest (e.g. BSgenome.Mmusculus.UCSC.mm10)</span>
<span class="kw">library</span>(BSgenome.Hsapiens.UCSC.hg38)</code></pre></div>
</div>
<div id="importing-data" class="section level2">
<h2>Importing data</h2>
<div id="genome-annotation" class="section level3">
<h3>Genome annotation</h3>
<p>We then import our gene and TE annotation files as <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html">GRanges objects</a> followed by Repexpr-specific curation and formatting using the <code>curateGenes</code> and <code>curateTEs</code> functions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># path to Gencode gtf file (provided)</span>
gene_path &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">'Repexpr'</span>, 
                        <span class="st">'extdata'</span>, 
                        <span class="st">'gencode.v29.annotation.gtf.gz'</span>)
<span class="co"># path to RepeatMasker hg38 repeat annotation (provided)</span>
rmsk_path &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">'Repexpr'</span>, 
                         <span class="st">'extdata'</span>, 
                         <span class="st">'hg38.fa.out.gz'</span>)
               
<span class="co"># import</span>
genes &lt;-<span class="st"> </span>rtracklayer<span class="op">::</span><span class="kw">import</span>(gene_path)
tes   &lt;-<span class="st"> </span><span class="kw">importRMSK</span>(rmsk_path)           <span class="co"># by default only imports DNA, SINE, LINE, and LTR class repeats on the main reference chromosomes                           </span>

<span class="co"># curation</span>
genes_cur &lt;-<span class="st"> </span><span class="kw">curateGenes</span>(genes,          <span class="co"># only retains exons, resolves naming ambiguity</span>
                         <span class="dt">extend =</span> <span class="dv">100</span>)   <span class="co"># extends exon coordinates 100 bps towards the 5'</span>
tes_cur   &lt;-<span class="st"> </span><span class="kw">curateTEs</span>(tes,              <span class="co"># tries to resolve nested TE intervals and adds annotation</span>
                       genes_cur)        <span class="co"># optional to flag tes overlapping exons)</span></code></pre></div>
</div>
<div id="reads" class="section level3">
<h3>Reads</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># path to bam files containing mapped reads                         </span>
bam_paths &lt;-<span class="st"> </span><span class="kw">dir</span>(<span class="st">'/net/mraid14/export/data/users/davidbr/proj/epitherapy/data/h1299/10x/dacsb/aligned/chunked_bams/'</span>,
                 <span class="dt">pattern =</span> <span class="st">'_deduplicated_chr[0-9, X, Y, MT]+.bam'</span>, 
                 <span class="dt">full.names =</span> <span class="ot">TRUE</span>)
bam_paths

reads &lt;-<span class="st"> </span><span class="kw">importBAM</span>(bam_paths, 
                   <span class="dt">paired =</span> <span class="ot">TRUE</span>,        <span class="co"># use FALSE for single-end libraries</span>
                   <span class="dt">mate =</span> <span class="st">'first'</span>,       <span class="co"># only imports the first mate of properly aligned read pairs, set to NA when using single-end libraries</span>
                   <span class="dt">anchor =</span> <span class="st">'fiveprime'</span>) <span class="co"># only retains 5' start of first mate as mapping position</span></code></pre></div>
</div>
</div>
<div id="compute-multiple-sequence-alignments" class="section level2">
<h2>Compute multiple sequence alignments</h2>
<p>Repexpr computes the read/UMI coverage along genes and the consensus model of TE families. This can be useful to sanity check 5’/3’ enrichment (depending on the protocol) and to identify putative TE consensus TSSs (5’ protocols), polyA-sites (3’ protocols), and to distinguish true de-repression from spurious background signal (e.g. intronic TE read mis-assignment, broad-scale genomic background transcription, etc.). As a rough estimate, we can utilize the consensus mapping information from Repeatmasker or DFAM output files for that purpose. This will usually provide reasonable results for highly conserved families. To increase accuracy, Repexpr can also compute family-wise multiple sequence alignments to improve mapping of individual loci onto a de novo alignment. When time and computational ressources are no limitation, we highly recommend this step by running:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="kw">compAlignments</span>(tes_cur, 
               <span class="dt">BSgenome  =</span> Hsapiens, <span class="co"># Genome to retrieve sequences</span>
               <span class="dt">max_gap   =</span> <span class="fl">0.95</span>,      <span class="co"># max gap percentage in the alignment to consider</span>
               <span class="dt">n_jobs    =</span> <span class="dv">500</span>,        <span class="co"># number of parallel jobs (needs HPC access)</span>
               <span class="dt">outdir    =</span> <span class="kw">tempdir</span>(),  <span class="co"># directory to write alignments and conversion table, ensure sufficient disk space!     </span>
               <span class="dt">overwrite =</span> <span class="ot">FALSE</span>)

## Compute count matrix

After we have imported and curated the data, we can procede to generate the actual read<span class="op">/</span>UMI count matrix using the <span class="st">`</span><span class="dt">compCounts</span><span class="st">`</span> function.
 </code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts &lt;-<span class="st"> </span><span class="kw">compCounts</span>(reads,                  <span class="co"># counts reads in TE/gene intervals and assigns them to their relative position on the consensus</span>
                     tes_cur,
                     genes_cur,
                     <span class="dt">bin_size =</span> <span class="dv">20</span>,
                     <span class="dt">conversion =</span> <span class="ot">NULL</span>,
                     <span class="dt">protocol   =</span> <span class="st">'fiveprime'</span>) <span class="co"># Suppresses extension of TE intervals into 3' genomic neighbourhood (useful for 3' based sequencing protocols)</span></code></pre></div>
</div>
<div id="call-cells" class="section level2">
<h2>Call cells</h2>
<p>To distinguish real cells from empty droplets, we utilize the <code>emptyDrops</code> function from the <a href="https://rdrr.io/github/MarioniLab/DropletUtils/">DropletUtils</a> package[1].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts_filt &lt;-<span class="st"> </span><span class="kw">callCells</span>(counts, <span class="dt">fdr =</span> <span class="fl">0.01</span>)</code></pre></div>
</div>
</div>
<div id="workflow-mouse-smart-seq2-scrna-seq-dataset" class="section level1">
<h1>Workflow mouse Smart-seq2 scRNA-seq dataset</h1>
<p>In this part of the tutorial, we will use Smart-seq2 data generated on early mouse embryos[1] to illustrate the specifics of full-coverage scRNA-seq data.</p>
<div id="getting-started-1" class="section level2">
<h2>Getting started</h2>
<p>We start the workflow by loading Repexpr and the mouse mm10 <em>BSgenome</em> object into our R environment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw">load_all</span>(<span class="st">'/net/mraid14/export/tgdata/users/davidbr/src/Repexpr/'</span>)

<span class="co"># adjust to your genome of interest (e.g. BSgenome.Hsapiens.UCSC.hg38)</span>
<span class="kw">library</span>(BSgenome.Mmusculus.UCSC.mm10)</code></pre></div>
</div>
<div id="import-genome-annotation" class="section level2">
<h2>Import genome annotation</h2>
<p>As before, we first import our gene and TE annotation files as <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html">GRanges objects</a> followed by Repexpr-specific curation and formatting using the <code>curateGenes</code> and <code>curateTEs</code> functions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># path to Gencode gtf file (provided)</span>
gene_path &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">'Repexpr'</span>, 
                        <span class="st">'extdata'</span>, 
                        <span class="st">'gencode.vM21.annotation.gtf.gz'</span>)
<span class="co"># path to RepeatMasker hg38 repeat annotation (provided)</span>
rmsk_path &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">'Repexpr'</span>, 
                         <span class="st">'extdata'</span>, 
                         <span class="st">'mm10.fa.out.gz'</span>)
                            
<span class="co"># import</span>
genes &lt;-<span class="st"> </span>rtracklayer<span class="op">::</span><span class="kw">import</span>(gene_path)
tes   &lt;-<span class="st"> </span><span class="kw">importRMSK</span>(rmsk_path)         <span class="co"># by default only imports DNA, SINE, LINE, and LTR class repeats on the main reference chromosomes                           </span>

<span class="co"># curation</span>
genes_cur &lt;-<span class="st"> </span><span class="kw">curateGenes</span>(genes)        <span class="co"># only retains exons, resolves naming ambiguity</span>
tes_cur   &lt;-<span class="st"> </span><span class="kw">curateTEs</span>(tes,            <span class="co"># tries to resolve nested TE intervals and adds annotation</span>
                       genes)          <span class="co"># optional to flag tes overlapping exons)</span></code></pre></div>
</div>
<div id="import-reads-and-compute-count-matrix-in-parallel" class="section level2">
<h2>Import reads and compute count matrix in parallel</h2>
<p>Given the size of the example Smart-seq2 dataset and memory limitations of most systems, we will compute single-cell gene/TE counts for subsets of the BAM files and combine the processed results afterwards. For enhanced speed performance, we will employ the parallel backend of the <a href="https://github.com/HenrikBengtsson/future">future package</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># path to bam files containing mapped reads                         </span>
bam_paths &lt;-<span class="st"> </span><span class="kw">dir</span>(<span class="st">'/net/mraid14/export/tgdata/users/evghenic/data/cheng2019/aligned/'</span>, 
                 <span class="dt">pattern =</span> <span class="st">'duprm.bam'</span>, 
                 <span class="dt">full.names =</span> <span class="ot">TRUE</span>)
bam_paths

<span class="co"># vector of cell barcodes </span>
barcodes &lt;-<span class="st"> </span><span class="kw">substring</span>(<span class="kw">basename</span>(bam_paths), <span class="dv">1</span>, <span class="dv">10</span>)

<span class="co"># commands for parallel execution</span>
cmds   &lt;-<span class="st"> </span><span class="kw">list</span>()
chunks &lt;-<span class="st"> </span><span class="kw">chunk</span>(bam_paths, <span class="dt">chunk_size =</span> <span class="dv">10</span>)

<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(chunks))
{
  bam_chunks     &lt;-<span class="st"> </span><span class="kw">glue</span>(<span class="st">&quot;c({glue_collapse(single_quote(bam_paths[chunks[[i]]]), sep = ',')})&quot;</span>)
  barcode_chunks &lt;-<span class="st"> </span><span class="kw">glue</span>(<span class="st">&quot;c({glue_collapse(single_quote(barcodes[chunks[[i]]]), sep = ',')})&quot;</span>)
 
  cmds[[i]] &lt;-<span class="st"> </span><span class="kw">glue</span>(<span class="st">&quot;Repexpr:::compCounts(reads = Repexpr:::importBAM({bam_chunks}, paired = FALSE, mate = NA, anchor = 'fiveprime', barcode = {barcode_chunks}),</span>
<span class="st">                     tes_cur,</span>
<span class="st">                     genes_cur,</span>
<span class="st">                     bin_size = 20,</span>
<span class="st">                     protocol = 'full')&quot;</span>)
}

<span class="co"># distribute commands</span>

<span class="co"># create new empty env and fill with relevant</span>
empty           &lt;-<span class="st"> </span><span class="kw">new.env</span>(<span class="dt">parent =</span> <span class="kw">emptyenv</span>())
empty<span class="op">$</span>tes_cur   &lt;-<span class="st"> </span>tes_cur
empty<span class="op">$</span>genes_cur &lt;-<span class="st"> </span>genes_cur
  
<span class="co"># distribute, compute and combine</span>
counts &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">gcluster.run3</span>(<span class="dt">command_list =</span> cmds,  
                <span class="dt">packages =</span> <span class="kw">c</span>(<span class="st">&quot;data.table&quot;</span>, <span class="st">&quot;plyranges&quot;</span>, <span class="st">&quot;base&quot;</span>, <span class="st">&quot;stats&quot;</span>), 
                <span class="dt">envir =</span> empty, 
                <span class="dt">collapse_results =</span> <span class="ot">TRUE</span>,
                <span class="dt">io_saturation =</span> <span class="ot">TRUE</span>)</code></pre></div>
</div>
<div id="normalize-counts" class="section level2">
<h2>Normalize counts</h2>
<p>The full transcript coverage of Smart-seq2 comes with a cost: Unlike most scRNA-seq protocols it lacks UMIs to quantify true molecule counts within a cell. To approximate molecule counts based on Smart-seq2 data, Repexpr samples <code>N</code> molecules per cell with empirical weights learned from the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts_norm &lt;-<span class="st"> </span><span class="kw">normCounts</span>(counts, 
                          <span class="dt">N =</span> <span class="dv">20000</span>) <span class="co"># number of molecules to sample per cell</span>

<span class="co"># Workflow mouse 10x scRNA-seq dataset (3')</span>

In this part of the tutorial, we will use <span class="dv">3</span><span class="st">' scRNA-seq data generated on mouse embryos[1] to illustrate the specifics of 3'</span> scRNA<span class="op">-</span>seq data.

## Getting started

We start the workflow by loading Repexpr and the mouse mm10 _BSgenome_ object into our R environment<span class="op">:</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw">load_all</span>(<span class="st">'/net/mraid14/export/tgdata/users/davidbr/src/Repexpr/'</span>)
<span class="kw">library</span>(BSgenome.Mmusculus.UCSC.mm10)</code></pre></div>
</div>
<div id="import-genome-annotation-1" class="section level2">
<h2>Import genome annotation</h2>
<p>As before, we first import our gene and TE annotation files as <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html">GRanges objects</a> followed by Repexpr-specific curation and formatting using the <code>curateGenes</code> and <code>curateTEs</code> functions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># path to Gencode gtf file (provided)</span>
gene_path &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">'Repexpr'</span>, 
                        <span class="st">'extdata'</span>, 
                        <span class="st">'gencode.vM21.annotation.gtf.gz'</span>)
<span class="co"># path to RepeatMasker hg38 repeat annotation (provided)</span>
rmsk_path &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">'Repexpr'</span>, 
                         <span class="st">'extdata'</span>, 
                         <span class="st">'mm10.fa.out.gz'</span>)
                            
<span class="co"># import</span>
genes &lt;-<span class="st"> </span>rtracklayer<span class="op">::</span><span class="kw">import</span>(gene_path)
tes   &lt;-<span class="st"> </span><span class="kw">importRMSK</span>(rmsk_path)         <span class="co"># by default only imports DNA, SINE, LINE, and LTR class repeats on the main reference chromosomes                           </span>

<span class="co"># curation</span>
genes_cur &lt;-<span class="st"> </span><span class="kw">curateGenes</span>(genes)        <span class="co"># only retains exons, resolves naming ambiguity</span>
tes_cur   &lt;-<span class="st"> </span><span class="kw">curateTEs</span>(tes,            <span class="co"># tries to resolve nested TE intervals and adds annotation</span>
                       genes)          <span class="co"># optional to flag tes overlapping exons)</span></code></pre></div>
</div>
<div id="import-reads-and-compute-count-matrix-in-parallel-1" class="section level2">
<h2>Import reads and compute count matrix in parallel</h2>
<p>Given the size of the example Smart-seq2 dataset and memory limitations of most systems, we will compute single-cell gene/TE counts for subsets of the BAM files and combine the processed results afterwards. For enhanced speed performance, we will employ the parallel backend of the <a href="https://github.com/HenrikBengtsson/future">future package</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># path to bam files containing mapped reads                         </span>
bam_paths &lt;-<span class="st"> </span><span class="kw">dir</span>(<span class="st">'/net/mraid14/export/tgdata/users/davidbr/data/goettgens_2019/aligned/'</span>, 
                 <span class="dt">pattern =</span> <span class="st">'bam.bam$'</span>,
                 <span class="dt">full.names =</span> <span class="ot">TRUE</span>)
bam_paths

<span class="co"># commands for parallel execution</span>
cmds   &lt;-<span class="st"> </span><span class="kw">list</span>()
chunks &lt;-<span class="st"> </span><span class="kw">chunk</span>(bam_paths, <span class="dt">chunk_size =</span> <span class="dv">1</span>)

<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(chunks))
{
  bam_chunks     &lt;-<span class="st"> </span><span class="kw">glue</span>(<span class="st">&quot;c({glue_collapse(single_quote(bam_paths[chunks[[i]]]), sep = ',')})&quot;</span>)
 
  cmds[[i]] &lt;-<span class="st"> </span><span class="kw">glue</span>(<span class="st">&quot;Repexpr:::compCounts(reads = Repexpr:::importBAM({bam_chunks}, paired = FALSE, mate = NA, anchor = 'fiveprime'),</span>
<span class="st">                     tes_cur,</span>
<span class="st">                     genes_cur,</span>
<span class="st">                     bin_size = 20,</span>
<span class="st">                     protocol = 'full')&quot;</span>)
}

<span class="co"># distribute commands</span>

<span class="co"># create new empty env and fill with relevant</span>
empty           &lt;-<span class="st"> </span><span class="kw">new.env</span>(<span class="dt">parent =</span> <span class="kw">emptyenv</span>())
empty<span class="op">$</span>tes_cur   &lt;-<span class="st"> </span>tes_cur
empty<span class="op">$</span>genes_cur &lt;-<span class="st"> </span>genes_cur
  
<span class="co"># distribute, compute and combine</span>
res_gclust &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">gcluster.run3</span>(<span class="dt">command_list =</span> cmds,  
                <span class="dt">packages =</span> <span class="kw">c</span>(<span class="st">&quot;data.table&quot;</span>, <span class="st">&quot;plyranges&quot;</span>, <span class="st">&quot;base&quot;</span>, <span class="st">&quot;stats&quot;</span>), 
                <span class="dt">envir =</span> empty, 
                <span class="dt">collapse_results =</span> <span class="ot">TRUE</span>,
                <span class="dt">io_saturation =</span> <span class="ot">TRUE</span>)</code></pre></div>
</div>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<p>[1]</p>
</div>
<div id="session-information" class="section level1">
<h1>Session information</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()
<span class="co">#&gt; R version 3.5.3 (2019-03-11)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: CentOS Linux 7 (Core)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /net/mraid14/export/data/users/eladch/tools/CO7/mkl/2018.3/compilers_and_libraries_2018.3.222/linux/mkl/lib/intel64_lin/libmkl_gf_lp64.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices datasets  utils     methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] tgstat_2.3.2 misha_4.0.4 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] compiler_3.5.3  magrittr_1.5    tools_3.5.3     htmltools_0.3.6</span>
<span class="co">#&gt;  [5] yaml_2.2.0      Rcpp_1.0.1      stringi_1.4.3   rmarkdown_1.12 </span>
<span class="co">#&gt;  [9] knitr_1.22      stringr_1.4.0   xfun_0.5        digest_0.6.19  </span>
<span class="co">#&gt; [13] evaluate_0.13</span></code></pre></div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
