<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Repsc vignette • Repsc</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Repsc vignette">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Repsc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="../articles/Prerequisites.html">
    <span class="fa fa-play-circle"></span>
     
    Prerequisites
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/5'scRNA_human.html">5' scRNA-seq of human cancer cells</a>
    </li>
    <li>
      <a href="../articles/5'scRNA_mouse.html">5' scRNA-seq of mouse ES cells</a>
    </li>
    <li>
      <a href="../articles/3'scRNA_mouse.html">3' scRNA-seq of mouse embryos</a>
    </li>
    <li>
      <a href="../articles/Smartseq2_mouse.html">Smartseq2 of mouse embryos</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-th-list"></span>
     
    Reference
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-news"></span>
     
    News
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://bitbucket.org/tanaylab/repsc/src/default/">
    <span class="fa fa-bitbucket-square"></span>
     
    Bitbucket
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Repsc vignette</h1>
                        <h4 class="author">David Brocks</h4>
            
            <h4 class="date">2019-09-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/tanaylab/repsc/blob/master/../../../../../net/mraid14/export/tgdata/users/davidbr/src/Repsc/vignettes/5" _vs_3><code>../../../../../net/mraid14/export/tgdata/users/davidbr/src/Repsc/vignettes/5'_vs_3'_comparison.Rmd</code></a></small>
      <div class="hidden name"><code>5'_vs_3'_comparison.Rmd</code></div>

    </div>

    
    
<div id="workflow-human-10x-scrna-seq-dataset-5" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow-human-10x-scrna-seq-dataset-5" class="anchor"></a>Workflow human 10x scRNA-seq dataset (5’)</h1>
<p>In this tutorial, we are going to utilize 5’ scRNA-seq data on epigenetically de-repressed cancer cell lines to quantify transposable element (TE) expression levels at single-cell and locus resolution. Following the workflow, you’ll learn the specifics of Repsc to adapt it to your single-cell dataset.</p>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h2>
<p>We start the workflow by loading Repsc and the human hg38 <em>BSgenome</em> object into our R environment:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/Sys.time.html">Sys.time</a></span>()</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co">#&gt; [1] "2019-09-22 14:50:44 IDT"</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3">devtools<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/load_all.html">load_all</a></span>(<span class="st">'/net/mraid14/export/tgdata/users/davidbr/src/Repsc/'</span>)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co">#&gt; Loading Repsc</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5">devtools<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/load_all.html">load_all</a></span>(<span class="st">'/net/mraid14/export/tgdata/users/davidbr/src/Reputils/'</span>)</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">#&gt; Loading Reputils</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7">devtools<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/load_all.html">load_all</a></span>(<span class="st">'/net/mraid14/export/tgdata/users/davidbr/src/Repdata/'</span>)</a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="co">#&gt; Loading Repdata</span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"></a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(BSgenome.Hsapiens.UCSC.hg38)</a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="co">#&gt; Loading required package: BSgenome</span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="co">#&gt; Loading required package: BiocGenerics</span></a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="co">#&gt; Loading required package: parallel</span></a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-15" data-line-number="15"><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></a>
<a class="sourceLine" id="cb1-16" data-line-number="16"><span class="co">#&gt; The following objects are masked from 'package:parallel':</span></a>
<a class="sourceLine" id="cb1-17" data-line-number="17"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-18" data-line-number="18"><span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ, clusterExport, clusterMap, parApply, parCapply, parLapply, parLapplyLB, parRapply, parSapply, parSapplyLB</span></a>
<a class="sourceLine" id="cb1-19" data-line-number="19"><span class="co">#&gt; The following objects are masked from 'package:Reputils':</span></a>
<a class="sourceLine" id="cb1-20" data-line-number="20"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-21" data-line-number="21"><span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ, clusterExport, clusterMap, combine, intersect, parApply, parCapply, parLapply, parLapplyLB, parRapply, parSapply,</span></a>
<a class="sourceLine" id="cb1-22" data-line-number="22"><span class="co">#&gt;     parSapplyLB, setdiff, union, which</span></a>
<a class="sourceLine" id="cb1-23" data-line-number="23"><span class="co">#&gt; The following objects are masked from 'package:Repsc':</span></a>
<a class="sourceLine" id="cb1-24" data-line-number="24"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-25" data-line-number="25"><span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ, clusterExport, clusterMap, combine, counts, intersect, parApply, parCapply, parLapply, parLapplyLB, parRapply,</span></a>
<a class="sourceLine" id="cb1-26" data-line-number="26"><span class="co">#&gt;     parSapply, parSapplyLB, setdiff, union, which</span></a>
<a class="sourceLine" id="cb1-27" data-line-number="27"><span class="co">#&gt; The following objects are masked from 'package:stats':</span></a>
<a class="sourceLine" id="cb1-28" data-line-number="28"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-29" data-line-number="29"><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></a>
<a class="sourceLine" id="cb1-30" data-line-number="30"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-31" data-line-number="31"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-32" data-line-number="32"><span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colMeans, colnames, colSums, dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep, grepl, intersect,</span></a>
<a class="sourceLine" id="cb1-33" data-line-number="33"><span class="co">#&gt;     is.unsorted, lapply, lengths, Map, mapply, match, mget, order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce, rowMeans, rownames, rowSums, sapply,</span></a>
<a class="sourceLine" id="cb1-34" data-line-number="34"><span class="co">#&gt;     setdiff, sort, table, tapply, union, unique, unsplit, which, which.max, which.min</span></a>
<a class="sourceLine" id="cb1-35" data-line-number="35"><span class="co">#&gt; Loading required package: S4Vectors</span></a>
<a class="sourceLine" id="cb1-36" data-line-number="36"><span class="co">#&gt; Loading required package: stats4</span></a>
<a class="sourceLine" id="cb1-37" data-line-number="37"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-38" data-line-number="38"><span class="co">#&gt; Attaching package: 'S4Vectors'</span></a>
<a class="sourceLine" id="cb1-39" data-line-number="39"><span class="co">#&gt; The following objects are masked from 'package:Reputils':</span></a>
<a class="sourceLine" id="cb1-40" data-line-number="40"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-41" data-line-number="41"><span class="co">#&gt;     first, rename, second, values</span></a>
<a class="sourceLine" id="cb1-42" data-line-number="42"><span class="co">#&gt; The following objects are masked from 'package:Repsc':</span></a>
<a class="sourceLine" id="cb1-43" data-line-number="43"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-44" data-line-number="44"><span class="co">#&gt;     first, rename, second, values</span></a>
<a class="sourceLine" id="cb1-45" data-line-number="45"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-46" data-line-number="46"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-47" data-line-number="47"><span class="co">#&gt;     expand.grid</span></a>
<a class="sourceLine" id="cb1-48" data-line-number="48"><span class="co">#&gt; Loading required package: IRanges</span></a>
<a class="sourceLine" id="cb1-49" data-line-number="49"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-50" data-line-number="50"><span class="co">#&gt; Attaching package: 'IRanges'</span></a>
<a class="sourceLine" id="cb1-51" data-line-number="51"><span class="co">#&gt; The following objects are masked from 'package:Reputils':</span></a>
<a class="sourceLine" id="cb1-52" data-line-number="52"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-53" data-line-number="53"><span class="co">#&gt;     collapse, desc, shift, slice, trim</span></a>
<a class="sourceLine" id="cb1-54" data-line-number="54"><span class="co">#&gt; The following objects are masked from 'package:Repsc':</span></a>
<a class="sourceLine" id="cb1-55" data-line-number="55"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-56" data-line-number="56"><span class="co">#&gt;     collapse, desc, shift, slice, trim</span></a>
<a class="sourceLine" id="cb1-57" data-line-number="57"><span class="co">#&gt; Loading required package: GenomeInfoDb</span></a>
<a class="sourceLine" id="cb1-58" data-line-number="58"><span class="co">#&gt; Loading required package: GenomicRanges</span></a>
<a class="sourceLine" id="cb1-59" data-line-number="59"><span class="co">#&gt; Loading required package: Biostrings</span></a>
<a class="sourceLine" id="cb1-60" data-line-number="60"><span class="co">#&gt; Loading required package: XVector</span></a>
<a class="sourceLine" id="cb1-61" data-line-number="61"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-62" data-line-number="62"><span class="co">#&gt; Attaching package: 'Biostrings'</span></a>
<a class="sourceLine" id="cb1-63" data-line-number="63"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-64" data-line-number="64"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-65" data-line-number="65"><span class="co">#&gt;     strsplit</span></a>
<a class="sourceLine" id="cb1-66" data-line-number="66"><span class="co">#&gt; Loading required package: rtracklayer</span></a>
<a class="sourceLine" id="cb1-67" data-line-number="67"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-68" data-line-number="68"><span class="co">#&gt; Attaching package: 'rtracklayer'</span></a>
<a class="sourceLine" id="cb1-69" data-line-number="69"><span class="co">#&gt; The following object is masked from 'package:Repsc':</span></a>
<a class="sourceLine" id="cb1-70" data-line-number="70"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-71" data-line-number="71"><span class="co">#&gt;     export</span></a></code></pre></div>
</div>
<div id="create-scset" class="section level2">
<h2 class="hasAnchor">
<a href="#create-scset" class="anchor"></a>Create scSet</h2>
<p>We then import our gene and TE annotation files as <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html">GRanges objects</a> followed by Repsc-specific curation and formatting using the <code>curateGenes</code> and <code>curateTEs</code> functions.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># path to Gencode gtf file (provided)</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="co"># gene_path &lt;- system.file(package = 'Repdata', </span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co">#                        'extdata', </span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co">#                        'gencode.v29.annotation.gtf.gz')</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5">gene_path &lt;-<span class="st"> '~/davidbr/tools/cellranger/references/refdata-cellranger-GRCh38-1.2.0/genes/genes.gtf'</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="co"># path to RepeatMasker hg38 repeat annotation (provided)</span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9">rmsk_path &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="dt">package =</span> <span class="st">'Repdata'</span>, </a>
<a class="sourceLine" id="cb2-10" data-line-number="10">                         <span class="st">'extdata'</span>, </a>
<a class="sourceLine" id="cb2-11" data-line-number="11">                         <span class="st">'hg38.fa.out.gz'</span>)</a>
<a class="sourceLine" id="cb2-12" data-line-number="12">                         </a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="co"># creating the scSet</span></a>
<a class="sourceLine" id="cb2-14" data-line-number="14">sc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/createScSet.html">createScSet</a></span>(<span class="dt">genome   =</span> Hsapiens,</a>
<a class="sourceLine" id="cb2-15" data-line-number="15">                  <span class="dt">protocol =</span> <span class="st">'threeprime'</span>,</a>
<a class="sourceLine" id="cb2-16" data-line-number="16">                  <span class="dt">tes      =</span> rmsk_path,</a>
<a class="sourceLine" id="cb2-17" data-line-number="17">                  <span class="dt">genes    =</span> gene_path)</a></code></pre></div>
</div>
<div id="compute-count-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#compute-count-matrix" class="anchor"></a>Compute count matrix</h2>
<p>After we have imported and curated the data, we can procede to generate the actual read/UMI count matrix using the <code>compCounts</code> function.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># path to bam files containing mapped reads                         </span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">bam_paths &lt;-<span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">                <span class="kw"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="st">'/net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/aligned/chunked_bams'</span>, </a>
<a class="sourceLine" id="cb3-4" data-line-number="4">                  <span class="dt">pattern =</span> <span class="st">'_deduplicated_chr[0-9, X, Y, MT]+.bam'</span>, </a>
<a class="sourceLine" id="cb3-5" data-line-number="5">                  <span class="dt">full.names =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb3-6" data-line-number="6">                <span class="kw"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="st">'/net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/parental/aligned/chunked_bams'</span>, </a>
<a class="sourceLine" id="cb3-7" data-line-number="7">                  <span class="dt">pattern =</span> <span class="st">'_deduplicated_chr[0-9, X, Y, MT]+.bam'</span>, </a>
<a class="sourceLine" id="cb3-8" data-line-number="8">                  <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb3-9" data-line-number="9">                )</a>
<a class="sourceLine" id="cb3-10" data-line-number="10"></a>
<a class="sourceLine" id="cb3-11" data-line-number="11">hdf5_paths &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(</a>
<a class="sourceLine" id="cb3-12" data-line-number="12">                <span class="st">'/net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/hct116_DMSO/outs/filtered_gene_bc_matrices_h5.h5'</span>,</a>
<a class="sourceLine" id="cb3-13" data-line-number="13">                <span class="st">'/net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/parental/hct116_PAR/outs/filtered_gene_bc_matrices_h5.h5'</span></a>
<a class="sourceLine" id="cb3-14" data-line-number="14">                )             </a>
<a class="sourceLine" id="cb3-15" data-line-number="15"></a>
<a class="sourceLine" id="cb3-16" data-line-number="16"><span class="co"># create a data.frame specifying import parameters                 </span></a>
<a class="sourceLine" id="cb3-17" data-line-number="17">bam_df    &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">paths   =</span> bam_paths,</a>
<a class="sourceLine" id="cb3-18" data-line-number="18">                        <span class="dt">paired  =</span> <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="ot">TRUE</span>, <span class="ot">FALSE</span>), <span class="dt">each =</span> <span class="dv">25</span>),       <span class="co"># use FALSE for single-end libraries</span></a>
<a class="sourceLine" id="cb3-19" data-line-number="19">                        <span class="dt">mate    =</span> <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'first'</span>, <span class="ot">NA</span>), <span class="dt">each =</span> <span class="dv">25</span>),    <span class="co"># only imports the first mate of properly aligned read pairs, set to NA when using single-end libraries</span></a>
<a class="sourceLine" id="cb3-20" data-line-number="20">                        <span class="dt">barcode =</span> <span class="st">'CB'</span>,       <span class="co"># 10x barcode included in BAM flag</span></a>
<a class="sourceLine" id="cb3-21" data-line-number="21">                        <span class="dt">chunk   =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">50</span>,</a>
<a class="sourceLine" id="cb3-22" data-line-number="22">                        <span class="dt">hdf5    =</span> <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(hdf5_paths, <span class="dt">each =</span> <span class="dv">25</span>),</a>
<a class="sourceLine" id="cb3-23" data-line-number="23">                        <span class="dt">meta    =</span> <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'fivep'</span>, <span class="st">'threep'</span>), <span class="dt">each =</span> <span class="dv">25</span>),</a>
<a class="sourceLine" id="cb3-24" data-line-number="24">                        <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb3-25" data-line-number="25">                        </a>
<a class="sourceLine" id="cb3-26" data-line-number="26"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(bam_df)                      </a>
<a class="sourceLine" id="cb3-27" data-line-number="27"><span class="co">#&gt;                                                                                                                               paths paired  mate barcode chunk</span></a>
<a class="sourceLine" id="cb3-28" data-line-number="28"><span class="co">#&gt; 1  /net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/aligned/chunked_bams/possorted_deduplicated_chr1.bam   TRUE first      CB     1</span></a>
<a class="sourceLine" id="cb3-29" data-line-number="29"><span class="co">#&gt; 2 /net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/aligned/chunked_bams/possorted_deduplicated_chr10.bam   TRUE first      CB     2</span></a>
<a class="sourceLine" id="cb3-30" data-line-number="30"><span class="co">#&gt; 3 /net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/aligned/chunked_bams/possorted_deduplicated_chr11.bam   TRUE first      CB     3</span></a>
<a class="sourceLine" id="cb3-31" data-line-number="31"><span class="co">#&gt; 4 /net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/aligned/chunked_bams/possorted_deduplicated_chr12.bam   TRUE first      CB     4</span></a>
<a class="sourceLine" id="cb3-32" data-line-number="32"><span class="co">#&gt; 5 /net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/aligned/chunked_bams/possorted_deduplicated_chr13.bam   TRUE first      CB     5</span></a>
<a class="sourceLine" id="cb3-33" data-line-number="33"><span class="co">#&gt; 6 /net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/aligned/chunked_bams/possorted_deduplicated_chr14.bam   TRUE first      CB     6</span></a>
<a class="sourceLine" id="cb3-34" data-line-number="34"><span class="co">#&gt;                                                                                                                           hdf5  meta</span></a>
<a class="sourceLine" id="cb3-35" data-line-number="35"><span class="co">#&gt; 1 /net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/hct116_DMSO/outs/filtered_gene_bc_matrices_h5.h5 fivep</span></a>
<a class="sourceLine" id="cb3-36" data-line-number="36"><span class="co">#&gt; 2 /net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/hct116_DMSO/outs/filtered_gene_bc_matrices_h5.h5 fivep</span></a>
<a class="sourceLine" id="cb3-37" data-line-number="37"><span class="co">#&gt; 3 /net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/hct116_DMSO/outs/filtered_gene_bc_matrices_h5.h5 fivep</span></a>
<a class="sourceLine" id="cb3-38" data-line-number="38"><span class="co">#&gt; 4 /net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/hct116_DMSO/outs/filtered_gene_bc_matrices_h5.h5 fivep</span></a>
<a class="sourceLine" id="cb3-39" data-line-number="39"><span class="co">#&gt; 5 /net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/hct116_DMSO/outs/filtered_gene_bc_matrices_h5.h5 fivep</span></a>
<a class="sourceLine" id="cb3-40" data-line-number="40"><span class="co">#&gt; 6 /net/mraid14/export/data/users/davidbr/proj/epitherapy/data/hct116/10x/dmso/hct116_DMSO/outs/filtered_gene_bc_matrices_h5.h5 fivep</span></a></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># compute single-cell counts</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">sc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addCounts.html">addCounts</a></span>(sc,</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">                <span class="dt">bams         =</span> bam_df,</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">                <span class="dt">bin_size     =</span> <span class="dv">25</span>,</a>
<a class="sourceLine" id="cb4-5" data-line-number="5">                <span class="dt">msa_dir      =</span> <span class="ot">NULL</span>,</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">                <span class="dt">use_gcluster =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-7" data-line-number="7"></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="kw"><a href="../reference/plotCells.html">plotCells</a></span>(sc)                </a></code></pre></div>
</div>
<div id="resolve-tegene-ambiguity" class="section level2">
<h2 class="hasAnchor">
<a href="#resolve-tegene-ambiguity" class="anchor"></a>Resolve TE/Gene ambiguity</h2>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">sc &lt;-<span class="st"> </span><span class="kw">resolveAmbiguity</span>(sc)</a></code></pre></div>
</div>
<div id="call-tsss" class="section level2">
<h2 class="hasAnchor">
<a href="#call-tsss" class="anchor"></a>Call TSSs</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">sc &lt;-<span class="st"> </span><span class="kw">callPeaks</span>(sc, </a>
<a class="sourceLine" id="cb6-2" data-line-number="2">                <span class="dt">neighb_bins =</span> <span class="dv">10</span>,</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">                <span class="dt">offset      =</span> <span class="dv">4</span>,</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">                <span class="dt">min_expr    =</span> <span class="dv">10</span>,</a>
<a class="sourceLine" id="cb6-5" data-line-number="5">                <span class="dt">fdr         =</span> <span class="fl">0.01</span>)</a></code></pre></div>
</div>
<div id="call-genete-stats" class="section level2">
<h2 class="hasAnchor">
<a href="#call-genete-stats" class="anchor"></a>Call gene/TE stats</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">sc &lt;-<span class="st"> </span><span class="kw">addStats</span>(sc,</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">               <span class="dt">min_expr =</span> <span class="dv">10</span>,</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">               <span class="dt">min_norm_var =</span> <span class="fl">0.5</span>)</a></code></pre></div>
</div>
<div id="compare-global-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#compare-global-counts" class="anchor"></a>Compare global counts</h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">counts      &lt;-<span class="st"> </span>Repsc<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/Repsc/man/counts.html">counts</a></span>(sc)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">counts_aggr &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="st">  </span>counts[, .(<span class="dt">n =</span> <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(n_all, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>), <span class="dt">n_uniq =</span> <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(n)), by =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'name'</span>, <span class="st">'barcode'</span>)</a>
<a class="sourceLine" id="cb8-4" data-line-number="4">    ][, n <span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(n)</a>
<a class="sourceLine" id="cb8-5" data-line-number="5">      ][<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(n_uniq <span class="op">&gt;</span><span class="st"> </span>n), n <span class="op">:</span><span class="er">=</span><span class="st"> </span>n_uniq]</a>
<a class="sourceLine" id="cb8-6" data-line-number="6"></a>
<a class="sourceLine" id="cb8-7" data-line-number="7">counts_ds   &lt;-<span class="st"> </span>Reputils<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/Reputils/man/downsamp.html">downsamp</a></span>(counts_aggr, <span class="dt">full =</span> <span class="ot">FALSE</span>, <span class="dt">i =</span> <span class="st">'name'</span>)</a>
<a class="sourceLine" id="cb8-8" data-line-number="8">counts_ds   &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/merge.html">merge</a></span>(counts_ds, <span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(counts[, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'name'</span>, <span class="st">'barcode'</span>, <span class="st">'meta'</span>, <span class="st">'type'</span>)]), <span class="dt">all.x =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb8-9" data-line-number="9"></a>
<a class="sourceLine" id="cb8-10" data-line-number="10">summarized &lt;-</a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="st">  </span>counts_ds[, .(<span class="dt">n =</span> <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(n_ds)), by =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'name'</span>, <span class="st">'type'</span>, <span class="st">'meta'</span>)] </a>
<a class="sourceLine" id="cb8-12" data-line-number="12"></a>
<a class="sourceLine" id="cb8-13" data-line-number="13">p_glob_comp &lt;-</a>
<a class="sourceLine" id="cb8-14" data-line-number="14"><span class="st">  </span>summarized <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-15" data-line-number="15"><span class="st">    </span>tidyr<span class="op">::</span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/spread.html">spread</a></span>(meta, n, <span class="dt">fill =</span> <span class="dv">0</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-16" data-line-number="16"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(fivep, threep)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-17" data-line-number="17"><span class="st">      </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="fl">0.1</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-18" data-line-number="18"><span class="st">      </span><span class="kw">scale_x_log10</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-19" data-line-number="19"><span class="st">      </span><span class="kw">scale_y_log10</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-20" data-line-number="20"><span class="st">      </span><span class="kw">facet_wrap</span>(<span class="op">~</span>type)</a>
<a class="sourceLine" id="cb8-21" data-line-number="21">      </a>
<a class="sourceLine" id="cb8-22" data-line-number="22">p_ltr12c_cov &lt;-</a>
<a class="sourceLine" id="cb8-23" data-line-number="23"><span class="st">  </span>counts[<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(name <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'LTR12C'</span>, <span class="st">'AluY'</span>, <span class="st">'MER51'</span>, <span class="st">'L1PA2'</span>, <span class="st">'LTR13'</span>, <span class="st">'THE1B'</span>, <span class="st">'LTR7'</span>)), .(<span class="dt">n =</span> <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(n_all)), by =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'meta'</span>, <span class="st">'pos_con'</span>, <span class="st">'name'</span>)][<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(meta <span class="op">==</span><span class="st"> 'threep'</span>), n <span class="op">:</span><span class="er">=</span><span class="st"> </span>n <span class="op">*</span><span class="st"> </span><span class="dv">-1</span>] <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-24" data-line-number="24"><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(pos_con, n)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-25" data-line-number="25"><span class="st">      </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">'identity'</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-26" data-line-number="26"><span class="st">      </span><span class="kw">facet_wrap</span>(<span class="op">~</span>name, <span class="dt">scales =</span> <span class="st">'free'</span>)</a>
<a class="sourceLine" id="cb8-27" data-line-number="27">  </a></code></pre></div>
</div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>[1]</p>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</a></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#workflow-human-10x-scrna-seq-dataset-5">Workflow human 10x scRNA-seq dataset (5’)</a><ul class="nav nav-pills nav-stacked">
<li><a href="#getting-started">Getting started</a></li>
      <li><a href="#create-scset">Create scSet</a></li>
      <li><a href="#compute-count-matrix">Compute count matrix</a></li>
      <li><a href="#resolve-tegene-ambiguity">Resolve TE/Gene ambiguity</a></li>
      <li><a href="#call-tsss">Call TSSs</a></li>
      <li><a href="#call-genete-stats">Call gene/TE stats</a></li>
      <li><a href="#compare-global-counts">Compare global counts</a></li>
      </ul>
</li>
      <li><a href="#references">References</a></li>
      <li><a href="#session-information">Session information</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://scholar.google.de/citations?user=ossdDj4AAAAJ&amp;hl=en&amp;oi=ao">David Brocks</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
