<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Repsc vignette • Repsc</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Repsc vignette">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Repsc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="../articles/Prerequisites.html">
    <span class="fa fa-play-circle"></span>
     
    Prerequisites
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/5'scRNA_human.html">5' scRNA-seq of human cancer cells</a>
    </li>
    <li>
      <a href="../articles/5'scRNA_mouse.html">5' scRNA-seq of mouse ES cells</a>
    </li>
    <li>
      <a href="../articles/3'scRNA_mouse.html">3' scRNA-seq of mouse embryos</a>
    </li>
    <li>
      <a href="../articles/Smartseq2_mouse.html">Smartseq2 of mouse embryos</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-th-list"></span>
     
    Reference
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-news"></span>
     
    News
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://bitbucket.org/tanaylab/repsc/src/default/">
    <span class="fa fa-bitbucket-square"></span>
     
    Bitbucket
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Repsc vignette</h1>
                        <h4 class="author">David Brocks</h4>
            
            <h4 class="date">2019-07-10</h4>
      
      
      <div class="hidden name"><code>5'scRNA_human.Rmd</code></div>

    </div>

    
    
<div id="workflow-human-10x-scrna-seq-dataset-5" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow-human-10x-scrna-seq-dataset-5" class="anchor"></a>Workflow human 10x scRNA-seq dataset (5’)</h1>
<p>In this tutorial, we are going to utilize 5’ scRNA-seq data on epigenetically de-repressed cancer cell lines to quantify transposable element (TE) expression levels at single-cell and locus resolution. Following the workflow, you’ll learn the specifics of Repsc to adapt it to your single-cell dataset.</p>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h2>
<p>We start the workflow by loading Repsc and the human hg38 <em>BSgenome</em> object into our R environment:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/load_all">load_all</a></span>(<span class="st">'/net/mraid14/export/tgdata/users/davidbr/src/Repsc/'</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co"># adjust to your genome of interest (e.g. BSgenome.Mmusculus.UCSC.mm10)</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(BSgenome.Hsapiens.UCSC.hg38)</a></code></pre></div>
</div>
<div id="create-scset" class="section level2">
<h2 class="hasAnchor">
<a href="#create-scset" class="anchor"></a>Create scSet</h2>
<p>We then import our gene and TE annotation files as <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html">GRanges objects</a> followed by Repsc-specific curation and formatting using the <code>curateGenes</code> and <code>curateTEs</code> functions.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># path to Gencode gtf file (provided)</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">gene_path &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="dt">package =</span> <span class="st">'Repsc'</span>, </a>
<a class="sourceLine" id="cb2-3" data-line-number="3">                        <span class="st">'extdata'</span>, </a>
<a class="sourceLine" id="cb2-4" data-line-number="4">                        <span class="st">'gencode.v29.annotation.gtf.gz'</span>)</a>
<a class="sourceLine" id="cb2-5" data-line-number="5"></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co"># path to RepeatMasker hg38 repeat annotation (provided)</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7">rmsk_path &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="dt">package =</span> <span class="st">'Repsc'</span>, </a>
<a class="sourceLine" id="cb2-8" data-line-number="8">                         <span class="st">'extdata'</span>, </a>
<a class="sourceLine" id="cb2-9" data-line-number="9">                         <span class="st">'hg38.fa.out.gz'</span>)</a>
<a class="sourceLine" id="cb2-10" data-line-number="10">                         </a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="co"># creating the scSet</span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12">sc &lt;-<span class="st"> </span><span class="kw">createScSet</span>(<span class="dt">genome   =</span> Hsapiens,</a>
<a class="sourceLine" id="cb2-13" data-line-number="13">                  <span class="dt">protocol =</span> <span class="st">'fiveprime'</span>,</a>
<a class="sourceLine" id="cb2-14" data-line-number="14">                  <span class="dt">tes      =</span> rmsk_path,</a>
<a class="sourceLine" id="cb2-15" data-line-number="15">                  <span class="dt">genes    =</span> gene_path)</a></code></pre></div>
</div>
<div id="compute-multiple-sequence-alignments" class="section level2">
<h2 class="hasAnchor">
<a href="#compute-multiple-sequence-alignments" class="anchor"></a>Compute multiple sequence alignments</h2>
<p>Repsc computes the read/UMI coverage along genes and the consensus model of TE families. This can be useful to sanity check 5’/3’ enrichment (depending on the protocol) and to identify putative TE consensus TSSs (5’ protocols), polyA-sites (3’ protocols), and to distinguish true de-repression from spurious background signal (e.g. intronic TE read mis-assignment, broad-scale genomic background transcription, etc.). As a rough estimate, we can utilize the consensus mapping information from Repeatmasker or DFAM output files for that purpose. This will usually provide reasonable results for highly conserved families. To increase accuracy, Repsc can also compute family-wise multiple sequence alignments to improve mapping of individual loci onto a de novo alignment. When time and computational ressources are no limitation, we recommend this step by running:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw">compAlignments</span>(sc, </a>
<a class="sourceLine" id="cb3-3" data-line-number="3">               <span class="dt">max_gap   =</span> <span class="fl">0.95</span>,       <span class="co"># max gap percentage in the alignment to consider</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4">               <span class="dt">n_jobs    =</span> <span class="dv">500</span>,        <span class="co"># number of parallel jobs (needs HPC access)</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5">               <span class="dt">outdir    =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/tempfile">tempdir</a></span>(),      </a>
<a class="sourceLine" id="cb3-6" data-line-number="6">               <span class="dt">overwrite =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb3-7" data-line-number="7"></a>
<a class="sourceLine" id="cb3-8" data-line-number="8">## Compute count matrix</a>
<a class="sourceLine" id="cb3-9" data-line-number="9"></a>
<a class="sourceLine" id="cb3-10" data-line-number="10">After we have imported and curated the data, we can procede to generate the actual read<span class="op">/</span>UMI count matrix using the <span class="st">`</span><span class="dt">compCounts</span><span class="st">`</span> function.</a>
<a class="sourceLine" id="cb3-11" data-line-number="11"> </a></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># path to bam files containing mapped reads                         </span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">bam_paths &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">dir</a></span>(<span class="st">'/net/mraid14/export/data/users/davidbr/proj/epitherapy/data/h1299/10x/dacsb/aligned/chunked_bams/'</span>,</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">                 <span class="dt">pattern =</span> <span class="st">'_deduplicated_chr[0-9, X, Y, MT]+.bam'</span>, </a>
<a class="sourceLine" id="cb4-4" data-line-number="4">                 <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5">                 </a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co"># create a data.frame specifying import parameters                 </span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7">bam_df    &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">paths   =</span> bam_paths,</a>
<a class="sourceLine" id="cb4-8" data-line-number="8">                        <span class="dt">paired  =</span> <span class="ot">TRUE</span>,       <span class="co"># use FALSE for single-end libraries</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9">                        <span class="dt">mate    =</span> <span class="st">'first'</span>,    <span class="co"># only imports the first mate of properly aligned read pairs, set to NA when using single-end libraries</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10">                        <span class="dt">barcode =</span> <span class="st">'CB'</span>,       <span class="co"># 10x barcode included in BAM flag</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11">                        <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb4-12" data-line-number="12">sc &lt;-<span class="st"> </span><span class="kw">addCounts</span>(sc,</a>
<a class="sourceLine" id="cb4-13" data-line-number="13">                <span class="dt">reads    =</span> bam_df[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>,],</a>
<a class="sourceLine" id="cb4-14" data-line-number="14">                <span class="dt">bin_size =</span> <span class="dv">25</span>,</a>
<a class="sourceLine" id="cb4-15" data-line-number="15">                <span class="dt">msa_dir  =</span> <span class="ot">NULL</span>)</a></code></pre></div>
</div>
<div id="call-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#call-cells" class="anchor"></a>Call cells</h2>
<p>To distinguish real cells from empty droplets, we utilize the <code>emptyDrops</code> function from the <a href="https://rdrr.io/github/MarioniLab/DropletUtils/">DropletUtils</a> package[1].</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">plotCellSize</span>(scSet)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"></a>
<a class="sourceLine" id="cb5-3" data-line-number="3">sc &lt;-<span class="st"> </span><span class="kw">callCells</span>(sc, </a>
<a class="sourceLine" id="cb5-4" data-line-number="4">                <span class="dt">method =</span> <span class="st">'size'</span>,</a>
<a class="sourceLine" id="cb5-5" data-line-number="5">                <span class="dt">min_size =</span> <span class="dv">1000</span>)</a></code></pre></div>
</div>
<div id="resolve-tegene-ambiguity" class="section level2">
<h2 class="hasAnchor">
<a href="#resolve-tegene-ambiguity" class="anchor"></a>Resolve TE/Gene ambiguity</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">sc &lt;-<span class="st"> </span><span class="kw">resolveAmbiguity</span>(sc)</a></code></pre></div>
</div>
<div id="call-tsss" class="section level2">
<h2 class="hasAnchor">
<a href="#call-tsss" class="anchor"></a>Call TSSs</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">test =<span class="st"> </span><span class="kw">callPeaks</span>(sc, </a>
<a class="sourceLine" id="cb7-2" data-line-number="2">          <span class="dt">neighb_bins =</span> <span class="dv">10</span>,</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">          <span class="dt">offset =</span> <span class="dv">4</span>,</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">          <span class="dt">min_umis =</span> <span class="dv">10</span>,</a>
<a class="sourceLine" id="cb7-5" data-line-number="5">          <span class="dt">fdr =</span> <span class="dv">0</span>)</a></code></pre></div>
</div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>[1]</p>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/sessionInfo">sessionInfo</a></span>()</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="co">#&gt; R version 3.5.3 (2019-03-11)</span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co">#&gt; Running under: CentOS Linux 7 (Core)</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-6" data-line-number="6"><span class="co">#&gt; Matrix products: default</span></a>
<a class="sourceLine" id="cb8-7" data-line-number="7"><span class="co">#&gt; BLAS/LAPACK: /net/mraid14/export/data/users/eladch/tools/CO7/mkl/2018.3/compilers_and_libraries_2018.3.222/linux/mkl/lib/intel64_lin/libmkl_gf_lp64.so</span></a>
<a class="sourceLine" id="cb8-8" data-line-number="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-9" data-line-number="9"><span class="co">#&gt; locale:</span></a>
<a class="sourceLine" id="cb8-10" data-line-number="10"><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></a>
<a class="sourceLine" id="cb8-12" data-line-number="12"><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></a>
<a class="sourceLine" id="cb8-13" data-line-number="13"><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></a>
<a class="sourceLine" id="cb8-14" data-line-number="14"><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></a>
<a class="sourceLine" id="cb8-15" data-line-number="15"><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></a>
<a class="sourceLine" id="cb8-16" data-line-number="16"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-17" data-line-number="17"><span class="co">#&gt; attached base packages:</span></a>
<a class="sourceLine" id="cb8-18" data-line-number="18"><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></a>
<a class="sourceLine" id="cb8-19" data-line-number="19"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb8-20" data-line-number="20"><span class="co">#&gt; loaded via a namespace (and not attached):</span></a>
<a class="sourceLine" id="cb8-21" data-line-number="21"><span class="co">#&gt;  [1] Rcpp_1.0.1       roxygen2_6.1.1   rprojroot_1.3-2  crayon_1.3.4    </span></a>
<a class="sourceLine" id="cb8-22" data-line-number="22"><span class="co">#&gt;  [5] assertthat_0.2.1 digest_0.6.19    commonmark_1.7   MASS_7.3-51.1   </span></a>
<a class="sourceLine" id="cb8-23" data-line-number="23"><span class="co">#&gt;  [9] R6_2.4.0         backports_1.1.4  magrittr_1.5     evaluate_0.13   </span></a>
<a class="sourceLine" id="cb8-24" data-line-number="24"><span class="co">#&gt; [13] stringi_1.4.3    rlang_0.3.2      fs_1.2.7         xml2_1.2.0      </span></a>
<a class="sourceLine" id="cb8-25" data-line-number="25"><span class="co">#&gt; [17] rmarkdown_1.12   pkgdown_1.3.0    desc_1.2.0       tools_3.5.3     </span></a>
<a class="sourceLine" id="cb8-26" data-line-number="26"><span class="co">#&gt; [21] stringr_1.4.0    yaml_2.2.0       xfun_0.5         compiler_3.5.3  </span></a>
<a class="sourceLine" id="cb8-27" data-line-number="27"><span class="co">#&gt; [25] memoise_1.1.0    htmltools_0.3.6  knitr_1.22</span></a></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#workflow-human-10x-scrna-seq-dataset-5">Workflow human 10x scRNA-seq dataset (5’)</a><ul class="nav nav-pills nav-stacked">
<li><a href="#getting-started">Getting started</a></li>
      <li><a href="#create-scset">Create scSet</a></li>
      <li><a href="#compute-multiple-sequence-alignments">Compute multiple sequence alignments</a></li>
      <li><a href="#call-cells">Call cells</a></li>
      <li><a href="#resolve-tegene-ambiguity">Resolve TE/Gene ambiguity</a></li>
      <li><a href="#call-tsss">Call TSSs</a></li>
      </ul>
</li>
      <li><a href="#references">References</a></li>
      <li><a href="#session-information">Session information</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://scholar.google.de/citations?user=ossdDj4AAAAJ&amp;hl=en&amp;oi=ao">David Brocks</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
