<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Repsc vignette â€¢ Repsc</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Repsc vignette">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Repsc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="../articles/Prerequisites.html">
    <span class="fa fa-play-circle"></span>
     
    Prerequisites
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/5'scRNA_human.html">5' scRNA-seq of human cancer cells</a>
    </li>
    <li>
      <a href="../articles/5'scRNA_mouse.html">5' scRNA-seq of mouse ES cells</a>
    </li>
    <li>
      <a href="../articles/3'scRNA_mouse.html">3' scRNA-seq of mouse embryos</a>
    </li>
    <li>
      <a href="../articles/Smartseq2_mouse.html">Smartseq2 of mouse embryos</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-th-list"></span>
     
    Reference
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-news"></span>
     
    News
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://bitbucket.org/tanaylab/repsc/src/default/">
    <span class="fa fa-bitbucket-square"></span>
     
    Bitbucket
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Repsc vignette</h1>
                        <h4 class="author">David Brocks</h4>
            
            <h4 class="date">2019-07-10</h4>
      
      
      <div class="hidden name"><code>Smartseq2_mouse.Rmd</code></div>

    </div>

    
    
<div id="workflow-mouse-smart-seq2-scrna-seq-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow-mouse-smart-seq2-scrna-seq-dataset" class="anchor"></a>Workflow mouse Smart-seq2 scRNA-seq dataset</h1>
<p>In this part of the tutorial, we will use Smart-seq2 data generated on early mouse embryos[1] to illustrate the specifics of full-coverage scRNA-seq data.</p>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h2>
<p>We start the workflow by loading Repsc and the mouse mm10 <em>BSgenome</em> object into our R environment:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/load_all">load_all</a></span>(<span class="st">'/net/mraid14/export/tgdata/users/davidbr/src/Repsc/'</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co"># adjust to your genome of interest (e.g. BSgenome.Hsapiens.UCSC.hg38)</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(BSgenome.Mmusculus.UCSC.mm10)</a></code></pre></div>
</div>
<div id="import-genome-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#import-genome-annotation" class="anchor"></a>Import genome annotation</h2>
<p>As before, we first import our gene and TE annotation files as <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html">GRanges objects</a> followed by Repsc-specific curation and formatting using the <code>curateGenes</code> and <code>curateTEs</code> functions.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># path to Gencode gtf file (provided)</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">gene_path &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="dt">package =</span> <span class="st">'Repsc'</span>, </a>
<a class="sourceLine" id="cb2-3" data-line-number="3">                        <span class="st">'extdata'</span>, </a>
<a class="sourceLine" id="cb2-4" data-line-number="4">                        <span class="st">'gencode.vM21.annotation.gtf.gz'</span>)</a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co"># path to RepeatMasker hg38 repeat annotation (provided)</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6">rmsk_path &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="dt">package =</span> <span class="st">'Repsc'</span>, </a>
<a class="sourceLine" id="cb2-7" data-line-number="7">                         <span class="st">'extdata'</span>, </a>
<a class="sourceLine" id="cb2-8" data-line-number="8">                         <span class="st">'mm10.fa.out.gz'</span>)</a>
<a class="sourceLine" id="cb2-9" data-line-number="9">                            </a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="co"># import</span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11">genes &lt;-<span class="st"> </span>rtracklayer<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/rtracklayer/topics/export">import</a></span>(gene_path)</a>
<a class="sourceLine" id="cb2-12" data-line-number="12">tes   &lt;-<span class="st"> </span><span class="kw">importRMSK</span>(rmsk_path)         <span class="co"># by default only imports DNA, SINE, LINE, and LTR class repeats on the main reference chromosomes                           </span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13"></a>
<a class="sourceLine" id="cb2-14" data-line-number="14"><span class="co"># curation</span></a>
<a class="sourceLine" id="cb2-15" data-line-number="15">genes_cur &lt;-<span class="st"> </span><span class="kw">curateGenes</span>(genes)        <span class="co"># only retains exons, resolves naming ambiguity</span></a>
<a class="sourceLine" id="cb2-16" data-line-number="16">tes_cur   &lt;-<span class="st"> </span><span class="kw">curateTEs</span>(tes,            <span class="co"># tries to resolve nested TE intervals and adds annotation</span></a>
<a class="sourceLine" id="cb2-17" data-line-number="17">                       genes_cur)      <span class="co"># to filter exonic TEs</span></a></code></pre></div>
</div>
<div id="import-reads-and-compute-count-matrix-in-parallel" class="section level2">
<h2 class="hasAnchor">
<a href="#import-reads-and-compute-count-matrix-in-parallel" class="anchor"></a>Import reads and compute count matrix in parallel</h2>
<p>Given the size of the example Smart-seq2 dataset and memory limitations of most systems, we will compute single-cell gene/TE counts for subsets of the BAM files and combine the processed results afterwards. For enhanced speed performance, we will employ the parallel backend of the <a href="https://github.com/HenrikBengtsson/future">future package</a>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># path to bam files containing mapped reads                         </span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">bam_paths &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">dir</a></span>(<span class="st">'/net/mraid14/export/tgdata/users/evghenic/data/cheng2019/aligned/'</span>, </a>
<a class="sourceLine" id="cb3-3" data-line-number="3">                 <span class="dt">pattern =</span> <span class="st">'duprm.bam'</span>, </a>
<a class="sourceLine" id="cb3-4" data-line-number="4">                 <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">bam_paths</a>
<a class="sourceLine" id="cb3-6" data-line-number="6"></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co"># vector of cell barcodes </span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8">barcodes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/substr">substring</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/basename">basename</a></span>(bam_paths), <span class="dv">1</span>, <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb3-9" data-line-number="9"></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="co"># commands for parallel execution</span></a>
<a class="sourceLine" id="cb3-11" data-line-number="11">cmds   &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>()</a>
<a class="sourceLine" id="cb3-12" data-line-number="12">chunks &lt;-<span class="st"> </span><span class="kw">chunk</span>(bam_paths, <span class="dt">chunk_size =</span> <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb3-13" data-line-number="13"></a>
<a class="sourceLine" id="cb3-14" data-line-number="14"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(chunks))</a>
<a class="sourceLine" id="cb3-15" data-line-number="15">{</a>
<a class="sourceLine" id="cb3-16" data-line-number="16">  bam_chunks     &lt;-<span class="st"> </span><span class="kw">glue</span>(<span class="st">"c({glue_collapse(single_quote(bam_paths[chunks[[i]]]), sep = ',')})"</span>)</a>
<a class="sourceLine" id="cb3-17" data-line-number="17">  barcode_chunks &lt;-<span class="st"> </span><span class="kw">glue</span>(<span class="st">"c({glue_collapse(single_quote(barcodes[chunks[[i]]]), sep = ',')})"</span>)</a>
<a class="sourceLine" id="cb3-18" data-line-number="18"> </a>
<a class="sourceLine" id="cb3-19" data-line-number="19">  cmds[[i]] &lt;-<span class="st"> </span><span class="kw">glue</span>(<span class="st">"Repsc:::compCounts(reads = Repsc:::importBAM({bam_chunks}, paired = FALSE, mate = NA, anchor = 'fiveprime', barcode = {barcode_chunks}),</span></a>
<a class="sourceLine" id="cb3-20" data-line-number="20"><span class="st">                     tes_cur,</span></a>
<a class="sourceLine" id="cb3-21" data-line-number="21"><span class="st">                     genes_cur,</span></a>
<a class="sourceLine" id="cb3-22" data-line-number="22"><span class="st">                     bin_size = 20,</span></a>
<a class="sourceLine" id="cb3-23" data-line-number="23"><span class="st">                     protocol = 'full')"</span>)</a>
<a class="sourceLine" id="cb3-24" data-line-number="24">}</a>
<a class="sourceLine" id="cb3-25" data-line-number="25"></a>
<a class="sourceLine" id="cb3-26" data-line-number="26"><span class="co"># distribute commands</span></a>
<a class="sourceLine" id="cb3-27" data-line-number="27"></a>
<a class="sourceLine" id="cb3-28" data-line-number="28"><span class="co"># create new empty env and fill with relevant</span></a>
<a class="sourceLine" id="cb3-29" data-line-number="29">empty           &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/environment">new.env</a></span>(<span class="dt">parent =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/environment">emptyenv</a></span>())</a>
<a class="sourceLine" id="cb3-30" data-line-number="30">empty<span class="op">$</span>tes_cur   &lt;-<span class="st"> </span>tes_cur</a>
<a class="sourceLine" id="cb3-31" data-line-number="31">empty<span class="op">$</span>genes_cur &lt;-<span class="st"> </span>genes_cur</a>
<a class="sourceLine" id="cb3-32" data-line-number="32">  </a>
<a class="sourceLine" id="cb3-33" data-line-number="33"><span class="co"># distribute, compute and combine</span></a>
<a class="sourceLine" id="cb3-34" data-line-number="34">counts &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb3-35" data-line-number="35"><span class="st">  </span><span class="kw">gcluster.run3</span>(<span class="dt">command_list =</span> cmds,  </a>
<a class="sourceLine" id="cb3-36" data-line-number="36">                <span class="dt">packages =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"data.table"</span>, <span class="st">"plyranges"</span>, <span class="st">"base"</span>, <span class="st">"stats"</span>), </a>
<a class="sourceLine" id="cb3-37" data-line-number="37">                <span class="dt">envir =</span> empty, </a>
<a class="sourceLine" id="cb3-38" data-line-number="38">                <span class="dt">collapse_results =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb3-39" data-line-number="39">                <span class="dt">io_saturation =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="normalize-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#normalize-counts" class="anchor"></a>Normalize counts</h2>
<p>The full transcript coverage of Smart-seq2 comes with a cost: Unlike most scRNA-seq protocols it lacks UMIs to quantify true molecule counts per cell. As an approximation, Repsc samples <code>N</code> molecules per cell with empirical weights learned from the data to simulate integer molecule counts.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">counts_norm &lt;-<span class="st"> </span><span class="kw">normCounts</span>(counts, </a>
<a class="sourceLine" id="cb4-2" data-line-number="2">                          <span class="dt">N =</span> <span class="dv">20000</span>) <span class="co"># number of molecules to sample per cell</span></a></code></pre></div>
</div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>[1]</p>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/sessionInfo">sessionInfo</a></span>()</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co">#&gt; R version 3.5.3 (2019-03-11)</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">#&gt; Running under: CentOS Linux 7 (Core)</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">#&gt; Matrix products: default</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">#&gt; BLAS/LAPACK: /net/mraid14/export/data/users/eladch/tools/CO7/mkl/2018.3/compilers_and_libraries_2018.3.222/linux/mkl/lib/intel64_lin/libmkl_gf_lp64.so</span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="co">#&gt; locale:</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></a>
<a class="sourceLine" id="cb5-14" data-line-number="14"><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="co">#&gt; attached base packages:</span></a>
<a class="sourceLine" id="cb5-18" data-line-number="18"><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></a>
<a class="sourceLine" id="cb5-19" data-line-number="19"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-20" data-line-number="20"><span class="co">#&gt; loaded via a namespace (and not attached):</span></a>
<a class="sourceLine" id="cb5-21" data-line-number="21"><span class="co">#&gt;  [1] Rcpp_1.0.1       roxygen2_6.1.1   rprojroot_1.3-2  crayon_1.3.4    </span></a>
<a class="sourceLine" id="cb5-22" data-line-number="22"><span class="co">#&gt;  [5] assertthat_0.2.1 digest_0.6.19    commonmark_1.7   MASS_7.3-51.1   </span></a>
<a class="sourceLine" id="cb5-23" data-line-number="23"><span class="co">#&gt;  [9] R6_2.4.0         backports_1.1.4  magrittr_1.5     evaluate_0.13   </span></a>
<a class="sourceLine" id="cb5-24" data-line-number="24"><span class="co">#&gt; [13] stringi_1.4.3    rlang_0.3.2      fs_1.2.7         xml2_1.2.0      </span></a>
<a class="sourceLine" id="cb5-25" data-line-number="25"><span class="co">#&gt; [17] rmarkdown_1.12   pkgdown_1.3.0    desc_1.2.0       tools_3.5.3     </span></a>
<a class="sourceLine" id="cb5-26" data-line-number="26"><span class="co">#&gt; [21] stringr_1.4.0    yaml_2.2.0       xfun_0.5         compiler_3.5.3  </span></a>
<a class="sourceLine" id="cb5-27" data-line-number="27"><span class="co">#&gt; [25] memoise_1.1.0    htmltools_0.3.6  knitr_1.22</span></a></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#workflow-mouse-smart-seq2-scrna-seq-dataset">Workflow mouse Smart-seq2 scRNA-seq dataset</a><ul class="nav nav-pills nav-stacked">
<li><a href="#getting-started">Getting started</a></li>
      <li><a href="#import-genome-annotation">Import genome annotation</a></li>
      <li><a href="#import-reads-and-compute-count-matrix-in-parallel">Import reads and compute count matrix in parallel</a></li>
      <li><a href="#normalize-counts">Normalize counts</a></li>
      </ul>
</li>
      <li><a href="#references">References</a></li>
      <li><a href="#session-information">Session information</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://scholar.google.de/citations?user=ossdDj4AAAAJ&amp;hl=en&amp;oi=ao">David Brocks</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
